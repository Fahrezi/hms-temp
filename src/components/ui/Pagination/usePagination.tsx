import { useEffect, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

const usePagination = () => {
  const location = useLocation();
  const navigate = useNavigate();

  const [page, setPage] = useState(() => {
    const searchParams = new URLSearchParams(location.search);
    return Number(searchParams.get('page')) || 1;
  });

  useEffect(() => {
    const searchParams = new URLSearchParams(location.search);
    const pageFromUrl = Number(searchParams.get('page'));

    if (!pageFromUrl || pageFromUrl !== page) {
      searchParams.set('page', page.toString());
      void navigate(`?${searchParams.toString()}`, { replace: true });
    }
  }, [page, location.search, navigate]);

  return { page, setPage };
};

export default usePagination;
