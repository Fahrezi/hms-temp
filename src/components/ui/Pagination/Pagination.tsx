import usePagination from './usePagination';

type PaginationProps = {
  maxPage: number;
};

const Pagination = (props: PaginationProps) => {
  const { maxPage } = props;
  const { page, setPage } = usePagination();

  const handler = (action: string) => {
    setPage((prev) => {
      const newPage = action === 'next' ? prev + 1 : prev - 1;
      return newPage;
    });
  };

  return (
    <div>
      {page > 1 && <button onClick={() => handler('prev')}>PREV</button>}
      {page < maxPage && <button onClick={() => handler('next')}>NEXT</button>}
    </div>
  );
};

export default Pagination;
