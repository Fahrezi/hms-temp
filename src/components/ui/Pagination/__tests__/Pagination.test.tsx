import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import Pagination from '@/components/ui/Pagination';

const mockSetPage = vi.fn();
let currentPage = 1;

vi.mock('@/components/ui/Pagination/usePagination', () => {
  return {
    default: () => ({
      page: currentPage,
      setPage: mockSetPage,
    }),
  };
});

describe('Pagination', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    currentPage = 1;
  });

  it('should render with no buttons when on page 1 of 1', () => {
    currentPage = 1;

    const { container } = render(<Pagination maxPage={1} />);

    expect(screen.queryByText('PREV')).not.toBeInTheDocument();
    expect(screen.queryByText('NEXT')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should only show NEXT button on first page', () => {
    currentPage = 1;

    render(<Pagination maxPage={3} />);

    expect(screen.queryByText('PREV')).not.toBeInTheDocument();
    expect(screen.getByText('NEXT')).toBeInTheDocument();
  });

  it('should show both buttons on middle page', () => {
    currentPage = 2;

    render(<Pagination maxPage={3} />);

    expect(screen.getByText('PREV')).toBeInTheDocument();
    expect(screen.getByText('NEXT')).toBeInTheDocument();
  });

  it('should only show PREV button on last page', () => {
    currentPage = 3;

    render(<Pagination maxPage={3} />);

    expect(screen.getByText('PREV')).toBeInTheDocument();
    expect(screen.queryByText('NEXT')).not.toBeInTheDocument();
  });

  it('should call setPage with next page when NEXT is clicked', () => {
    currentPage = 1;

    render(<Pagination maxPage={3} />);

    fireEvent.click(screen.getByText('NEXT'));

    expect(mockSetPage).toHaveBeenCalled();

    const setPageCallback = mockSetPage.mock.calls[0][0];

    const result = setPageCallback(1);

    expect(result).toBe(2);
  });

  it('should call setPage with previous page when PREV is clicked', () => {
    currentPage = 2;

    render(<Pagination maxPage={3} />);
    fireEvent.click(screen.getByText('PREV'));

    expect(mockSetPage).toHaveBeenCalled();

    const setPageCallback = mockSetPage.mock.calls[0][0];

    const result = setPageCallback(2);

    expect(result).toBe(1);
  });
});
