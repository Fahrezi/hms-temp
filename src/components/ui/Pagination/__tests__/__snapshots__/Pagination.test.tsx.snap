// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > should render with no buttons when on page 1 of 1 1`] = `
<div>
  <div />
</div>
`;
